testcases/kernel/syscalls/getrandom/getrandom02.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2015 Cedric Hnyda ced.hnyda@gmail.com
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it would be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * Further, this software is distributed without any warranty that it is
+ * free of the rightful claim of any third person regarding infringement
+ * or the like.  Any license provided herein, whether implied or
+ * otherwise, applies only to this software file.  Patent licenses, if
+ * any, provided herein do not apply to combinations of this program with
+ * other software, or any other product whatsoever.
+ *
+ */
+
+/*
+ * AUTHOR   : Cédric Hnyda
+ * DATE STARTED : 06/13/2015
+ *
+ *  Calls getrandom(2) with a NULL buffer and expects failure.
+ *
+ */
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/fcntl.h>
+#include <sys/syscall.h>
+#include <errno.h>
+#include <linux/random.h>
+#include <stdio.h>
+
+#include "test.h"
+#include "safe_macros.h"
+
+static int modes[] = {0, GRND_RANDOM, GRND_NONBLOCK,
+						GRND_RANDOM | GRND_NONBLOCK};
+
+char *TCID = "getrandom02";
+int TST_TOTAL = ARRAY_SIZE(modes);
+
+static unsigned char buf[256];
+static size_t size = 256;
+
+static void fill(void);
+static int check_content(int nb);
+
+int main(int ac, char **av)
+{
+	int lc, i;
+
+	tst_parse_opts(ac, av, NULL, NULL);
+
+	for (lc = 0; TEST_LOOPING(lc); lc++) {
+		tst_count = 0;
+
+		for i = 0; i < TST_TOTAL; i++) {
+			fill();
+			TEST(syscall(SYS_getrandom, buf, size, modes[i]));
```

Comment:
In the NONBLOCK mode we may get ret -1 with errno ```EAGAIN``` if there wasn't enough randomness in the system and the test will fail the test. (should happen if we executed the test repeatedly)

We should do something like:

```
        do {
                TEST(syscall(SYS_getrandom, buf, size, modes[i]));
        } while ((modes[i] & GRND_NONBLOCK) && TEST_RETURN == -1 && TEST_ERRNO == EAGAIN)
```

Also if you use ```TEST_RETURN``` you should use ```TEST_ERRNO``` instead of ```errno``` (the test ```TEST_ERRNO``` is ```errno``` stored after the call, the ```errno``` may be changed meanwhile by calling other functions from glibc.